### ISSUE TITLE
v1.14.2 Google Auth - Error 400(Not Found): The requested URL was not found on this server.

### ISSUE THREAD

**Issue by zeroum** [keeweb/keeweb#1499]
**Describe the bug**
With KeeWeb Desktop v1.14.2, installed from "KeeWeb-1.14.2.linux.x64.deb", when trying to sync with the Google Drive, it opens a new window in the browser, with the title "Error 400 (Not Found)!!1", and the error content "The requested URL was not found on this server."

It works when I use the Online Web App at https://app.keeweb.info/ (KeeWeb v1.14.2 (03c44d7, 2020-05-04), Environment: web, 
User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:75.0) Gecko/20100101 Firefox/75.0) 

Note: I am logged on 3 Google Accounts In the browser . The database is on the Google Drive of the second account. I don't have databases on other accounts. The order of the accounts has always been the same for years.

**To Reproduce**
1. With a Google Drive sync configured, click on Settings
2. Click in KeePass in the left menu
3. Click Sync
4. It will open a new window in your prefered browser with the not found error.

**Expected behavior**
It should present the Google Authentication page.

**Screenshots**
![image](https://user-images.githubusercontent.com/145057/81441070-7d5a5900-9147-11ea-814b-af52876613f6.png)

URL KeeWeb is trying to reach: https://accounts.google.com/signin/oauth/legacy/consent?authuser=0&part=AJi8hAM7SDfkZbPIiFRqdUSo_OEMc5bdP916uZ3Lf8Va2B1xnvH_TxWtEM_Kq_5Oksy1qHXoRls8zU7Ceod3Lw-W_pnvqHCC7tG4qxDc1Ltwn1P8GbzWx0_127cV2qvnx6VkqvvaBI5hw48-zaDuQ8uK5o1Wz2H2KMNI8Q86vfM3jf9CwiP0D6UFoV0Ps-XhL1UwI6Cgm19rnbgLXALm3vzDjIgsXUdbUhjwzo6uQ3ffes3O-FP2JZqLOaBpigVXpeKUOVSYdLpXJhEyv4OkqaRLkO1ZgdImEeT_2I5xSF93NwoWNVrJ---c0zkCTsu2exEYf8RuW7HFWHboJOZNpJk3mNqWu_if0RugOyr5JlSv4x3N5kdYOC6Pt3WBb1fu4dkmZJLsJa4OyKA0a-B18jM0ant4DxP82RnP9bsnqH7p5iyhZDbyj3kFeG2pOnI5azeGAUCJVlGl&as=Koey9ZYza88lj2smckAS5g&pli=1#

**Environment**
KeeWeb v1.14.2 (03c44d7, 2020-05-04)
Environment: electron v8.2.5
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) KeeWeb/1.14.2 Chrome/80.0.3987.165 Electron/8.2.5 Safari/537.36

Installed from: https://github.com/keeweb/keeweb/releases/download/v1.14.2/KeeWeb-1.14.2.linux.x64.deb

Running on Kubuntu 18.04, external browser:
Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:75.0) Gecko/20100101 Firefox/75.0

**Logs**
```
[INFO ] 2020-05-08T18:57:13.336Z [start-profiler] Started in 442ms: fetching=126ms, pre-init=161ms, loading modules=55ms, document ready=10ms, creating app model=2ms, loading configs=23ms, initializing modules=1ms, loading remote config=1ms, checking features=0ms, preparing app model=0ms, first view rendering=189ms, app ready event=0ms. Electron app started in 1131ms: loading main requires=2ms, getting data dir=0ms, checking for new version=0ms, getting asar file time=0ms, loading app requires=1ms, single instance lock=0ms, defining args=0ms, setting env=0ms, restoring preferences=1ms, reading app settings=0ms, app on ready=16ms, setting app options=0ms, setting system appearance=0ms, creating main window=27ms, setting menu=3ms, configuring main window=7ms, restoring main window position=1ms, setting global shortcuts=2ms, subscribing to power events=0ms, deleting old temp files=1ms, setting request handlers=0ms, main window ready=215ms, main window shown=5ms.
[INFO ] 2020-05-08T18:57:13.336Z [updater] Next update check will happen in 85869s
[INFO ] 2020-05-08T18:57:19.651Z [open:KeePass] File open request
[INFO ] 2020-05-08T18:57:19.651Z [open:KeePass] Open file from cache, will sync after load gdrive
[DEBUG] 2020-05-08T18:57:19.651Z [storage-cache] Load 323c7efd-530a-f6fa-c78d-64aeb8a59d8e
[DEBUG] 2020-05-08T18:57:19.654Z [storage-cache] Loaded 323c7efd-530a-f6fa-c78d-64aeb8a59d8e 1ms
[INFO ] 2020-05-08T18:57:19.654Z [open:KeePass] Loaded file from cache null
[INFO ] 2020-05-08T18:57:19.781Z [file] Opened file KeePass: 127ms, 6000 rounds, 38 kB
[DEBUG] 2020-05-08T18:57:19.781Z [app] Add last open file 323c7efd-530a-f6fa-c78d-64aeb8a59d8e KeePass gdrive 0B0-AVHW-cvWDU09QV0FfTDVDdXc 0B0-AVHW-cvWDb0prK3FkbVJmbE91SGN1b3FiV1ZidzhKbE9nPQ
[INFO ] 2020-05-08T18:57:19.855Z [open:KeePass] Sync just opened file
[INFO ] 2020-05-08T18:57:19.865Z [sync:KeePass] Sync started gdrive 0B0-AVHW-cvWDU09QV0FfTDVDdXc {}
[INFO ] 2020-05-08T18:57:19.866Z [sync:KeePass] Stat file
[INFO ] 2020-05-08T18:57:19.873Z [storage-oauth-listener] Starting OAuth listener on port 48149...
[INFO ] 2020-05-08T18:57:19.887Z [sync:KeePass] Stat error, not dirty browser-auth-started
[INFO ] 2020-05-08T18:57:19.887Z [sync:KeePass] Sync finished browser-auth-started
```

Network tab in KeeWeb Desktop dev tools does not show activity.

**Comment by antelle** [comment]
Hi! What happens if you navigate to this url?
`
https://accounts.google.com/o/oauth2/v2/auth?client_id=847548101761-h2pcl2p6m1tssnlqm0vrm33crlveccbr.apps.googleusercontent.com&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive&redirect_uri=http%3A%2F%2Flocalhost%3A48149%2Foauth-result%2Fgdrive.html&response_type=code&code_challenge=bf6sg2lFDD3yxt2Zb_nfqfk_K9xDip5C1oNTGD54fs4&code_challenge_method=S256
`

**Comment by antelle** [comment]
It doesn't happen for me, I think this error should be reported to google.
https://stackoverflow.com/questions/61401613/google-auth-endpoint-redirects-to-a-400-error-page

**Comment by zeroum** [comment]
> Hi! What happens if you navigate to this url?
> `https://accounts.google.com/o/oauth2/v2/auth?client_id=847548101761-h2pcl2p6m1tssnlqm0vrm33crlveccbr.apps.googleusercontent.com&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive&redirect_uri=http%3A%2F%2Flocalhost%3A48149%2Foauth-result%2Fgdrive.html&response_type=code&code_challenge=bf6sg2lFDD3yxt2Zb_nfqfk_K9xDip5C1oNTGD54fs4&code_challenge_method=S256`

Hello!

It goes to the account choice page successfully, then when I choose the account, it tries to redirect me to http://localhost:48149/oauth-result/gdrive.html?code=4%2FzgEN-HYpoK5_rJZFI497TftJbKp0PlkhtwXt1s-yOjyuo90nUACZo98mHBQ9gAeSb6IuBwAv5ZCzpBzTIx31qCo&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive

**Comment by antelle** [comment]
Interesting! The only difference is `state` parameter. Now, if you add some `state` there?
`https://accounts.google.com/o/oauth2/v2/auth?client_id=847548101761-h2pcl2p6m1tssnlqm0vrm33crlveccbr.apps.googleusercontent.com&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive&state=fa380bd0f104e8e98094bb7698f32df860ae584f8ab2b57426d3ba1d9dc755959eb164b0a10844e5bb89b24cba75ff18eaf1001a42ccf6c00cbfa7f9ab515646&redirect_uri=http%3A%2F%2Flocalhost%3A48149%2Foauth-result%2Fgdrive.html&response_type=code&code_challenge=pOvUS8mjIguvTKpMoVwbSvFbus6BzQn2OmF6a0_uPPE&code_challenge_method=S256`

**Comment by zeroum** [comment]
> Interesting! The only difference is `state` parameter. Now, if you add some `state` there?
> `https://accounts.google.com/o/oauth2/v2/auth?client_id=847548101761-h2pcl2p6m1tssnlqm0vrm33crlveccbr.apps.googleusercontent.com&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive&state=fa380bd0f104e8e98094bb7698f32df860ae584f8ab2b57426d3ba1d9dc755959eb164b0a10844e5bb89b24cba75ff18eaf1001a42ccf6c00cbfa7f9ab515646&redirect_uri=http%3A%2F%2Flocalhost%3A48149%2Foauth-result%2Fgdrive.html&response_type=code&code_challenge=pOvUS8mjIguvTKpMoVwbSvFbus6BzQn2OmF6a0_uPPE&code_challenge_method=S256`

This one is fine too :face_with_head_bandage:

**Comment by zeroum** [comment]
Is there a way to see the request or the url when it sends? I didn't find anything in keeweb's dev tools network tab.

**Comment by antelle** [comment]
Hm, that's really strange, now I wonder what's the difference. To see the url, open DevTools from settings / general / advanced, and paste this to Console:
```js
require('electron').shell.openExternal = arg => console.log(arg)
```

**Comment by zeroum** [comment]
Yeah, it spat the url https://accounts.google.com/o/oauth2/v2/auth?client_id=847548101761-h2pcl2p6m1tssnlqm0vrm33crlveccbr.apps.googleusercontent.com&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive&state=cf53fb613ca3b08a040b90013da668d6e20aa8ad1f8507024588e051367617b61eea52b985affc1b359a457253432cb4eaa266e0e4c3bfb06b22f2f32b5bd32a&redirect_uri=http%3A%2F%2Flocalhost%3A48149%2Foauth-result%2Fgdrive.html&response_type=code&code_challenge=JaNpmNN33B5oXhb8ESFbt95QGoHVBKpDSWqavF89tTs&code_challenge_method=S256

And this is working fine.

I've got a theory: When I open a link from an app inside the desktop, KDE captures the link before it sends it to the default browser, and it is somehow damaging the data on the way. That man in the middle is not being transparent. Now to debug that...

**Comment by antelle** [comment]
Even more interesting. Do you have a debugging proxy that can capture requests sent by browser? What happens if you for example set Chrome as default browser?

**Comment by zeroum** [comment]
Yep. that was it

![image](https://user-images.githubusercontent.com/145057/81450016-ea75ea80-9157-11ea-86b3-434ca464344d.png)

It was erroring on me when the option "in the application based on the contents of the URL" was checked.
I now checked the "in the following application" option, and it now works perfectly.

**Comment by antelle** [comment]
Oh, so it captured the contents of the url, and then, since it was a one-time url, it was not valid in browser anymore. That explains it!

**Comment by antelle** [comment]
I'm not sure if we'll ever come to in, since Electron doesn't provide a way to get a system-default browser, and standard open commands won't work with this setting either. So, in this case, we need to launch exactly the default browser with this url, without fetching its contents.

**Comment by zeroum** [comment]
No idea if it contacts Google during a preflight check. This KDE Plasma LTS version (v.5.12) is more than 2 years old. I'll see later if the problem still happens in the current version (v5.18).

**Comment by grossetti** [comment]
I have the same issue using Ubuntu 19.10 and KeeWeb 1.14.2. I tried to change the default browser from Firefox to Chrome but this does not change anything.

**Comment by PierreOcean** [comment]
I am having the same issue here. I was just closing the popup window but now realize the database is not synchronizing with GG Drive thus not between my MacOS and my Android phone.

**Comment by zeroum** [comment]
@PierreOcean Your problem seems to be related to another ticket, #1497, which is solved by updating KeeWeb to a newer version (as of now the latest one is v1.14.3).

**Comment by PierreOcean** [comment]
@zeroum True! Thanks! I thought it was up-to-date but it seems it was downloading the update, restarting KeeWeb and never installing the update.
It seems fixed though now it is saving my database in the wrong folder of GG Drive.
Any clue how to fix the path?

**Comment by gbatson** [comment]
Thanks. That worked perfectly. It appears you have the only keepass client that works on linux with gdrive (out of the box) Kudos!
